projects/s-rxjs-utils/src/lib/map-with-caching.ts

```diff
@@ -0,0 +1,46 @@
+import { ObjectWith } from "micro-dash";
+import { map } from "rxjs/operators";
+
+/**
+ * Applies a given function to each value in the array emitted by the source Observable, and emits the resulting array as an Observable. Each item in the array is cached using key generated by the given function so that the resulting arrays contain references to the exact same objects included in the previous emission, when possible.
+ *
+ * @param buildCacheKey A function that converts an upstream object into a string to use as the cache key. Needs to return a unique key for each item in the source array.
+ * @param buildDownstreamItem A function that converts an upstream object into a downstream object
+ *
+ * ```
+ * const mapWithCaching = mapArrayWithCaching(
+ *   (item) => item.toString(),
+ *   (item) => item + 1
+ * )
+ *
+ * source:         -[1, 2]---[1, 2, 3]---[2]--|
+ * mapWithCaching: -[2, 3]---[2, 3, 4]---[3]--|
+ * ```
+ */
+export function mapArrayWithCaching<U, D>(
```

Comment:
`[green]`
`createPipeable()` spells these out as `UpstreamType` and `DownstreamType`. Might be nice to be that explicit, and consistent.